azurerm/internal/services/compute/resource_arm_managed_disk.go

```diff
@@ -281,28 +282,212 @@ func resourceArmManagedDiskCreateUpdate(d *schema.ResourceData, meta interface{}
 }
 
+func resourceArmManagedDiskUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Compute.DisksClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	log.Printf("[INFO] preparing arguments for Azure ARM Managed Disk update.")
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+	storageAccountType := d.Get("storage_account_type").(string)
+
+	disk, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(disk.Response) {
+			return fmt.Errorf("Error Managed Disk %q (Resource Group %q) was not found", name, resourceGroup)
+		}
+
+		return fmt.Errorf("Error making Read request on Azure Managed Disk %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	diskUpdate := compute.DiskUpdate{
+		DiskUpdateProperties: &compute.DiskUpdateProperties{},
+	}
+
+	if d.HasChange("tags") {
+		t := d.Get("tags").(map[string]interface{})
+		diskUpdate.Tags = tags.Expand(t)
+	}
+
+	if d.HasChange("storage_account_type") {
+		var skuName compute.DiskStorageAccountTypes
+		for _, v := range compute.PossibleDiskStorageAccountTypesValues() {
+			if strings.EqualFold(storageAccountType, string(v)) {
+				skuName = v
+			}
+		}
+		diskUpdate.Sku = &compute.DiskSku{
+			Name: skuName,
+		}
+	}
+
+	if strings.EqualFold(storageAccountType, string(compute.UltraSSDLRS)) {
+		if d.HasChange("disk_iops_read_write") {
+			v := d.Get("disk_iops_read_write")
+			diskIOPS := int64(v.(int))
+			diskUpdate.DiskIOPSReadWrite = &diskIOPS
+		}
+
+		if d.HasChange("disk_mbps_read_write") {
+			v := d.Get("disk_mbps_read_write")
+			diskMBps := int32(v.(int))
+			diskUpdate.DiskMBpsReadWrite = &diskMBps
+		}
+	} else {
+		if d.HasChange("disk_iops_read_write") || d.HasChange("disk_mbps_read_write") {
+			return fmt.Errorf("[ERROR] disk_iops_read_write and disk_mbps_read_write are only available for UltraSSD disks")
+		}
+	}
+
+	if d.HasChange("os_type") {
+		diskUpdate.DiskUpdateProperties.OsType = compute.OperatingSystemTypes(d.Get("os_type").(string))
+	}
+
+	if d.HasChange("disk_size_gb") {
+		if old, new := d.GetChange("disk_size_gb"); new.(int) > old.(int) {
+			diskUpdate.DiskUpdateProperties.DiskSizeGB = utils.Int32(int32(new.(int)))
+		} else {
+			return fmt.Errorf("Error - New size must be greater than original size. Shrinking disks is not supported on Azure")
+		}
+	}
+
+	// if we are attached to a VM we bring down the VM as necessary for the operations which are not allowed while it's online
+	if disk.ManagedBy != nil {
+		virtualMachine, err := ParseVirtualMachineID(*disk.ManagedBy)
+		if err != nil {
+			return fmt.Errorf("Error parsing VMID %q for disk attachment: %+v", *disk.ManagedBy, err)
+		}
+		// check instanceView State
+		vmClient := meta.(*clients.Client).Compute.VMClient
+
+		instanceView, err := vmClient.InstanceView(ctx, virtualMachine.ResourceGroup, virtualMachine.Name)
```

Comment:
we need to lock on the Virtual Machine resource group/name here before doing these manipulations